contrib/adaptive-compression/adapt.c

```diff
@@ -0,0 +1,591 @@
+    unsigned const writeWaiting = ctx->adaptParam < ctx->stats.compressedCounter ? 1 : 0;
+    unsigned const createWaiting = ctx->adaptParam < ctx->stats.writeCounter ? 1 : 0;
+    unsigned const writeSlow = ((compressWaiting && createWaiting) || (createWaiting && !writeWaiting)) ? 1 : 0;
+    unsigned const compressSlow = ((writeWaiting && createWaiting) || (writeWaiting && !compressWaiting)) ? 1 : 0;
+    unsigned const createSlow = ((compressWaiting && writeWaiting) || (compressWaiting && !createWaiting)) ? 1 : 0;
+    DEBUGLOG(2, "ready: %u compressed: %u write: %u\n", ctx->stats.readyCounter, ctx->stats.compressedCounter, ctx->stats.writeCounter);
+    if (allSlow) {
+        reset = 1;
+    }
+    else if ((writeSlow || createSlow) && ctx->compressionLevel < (unsigned)ZSTD_maxCLevel()) {
+        DEBUGLOG(2, "increasing compression level %u\n", ctx->compressionLevel);
+        ctx->compressionLevel++;
+        reset = 1;
+    }
+    else if (compressSlow && ctx->compressionLevel > 1) {
+        DEBUGLOG(2, "decreasing compression level %u\n", ctx->compressionLevel);
+        ctx->compressionLevel--;
+        reset = 1;
+    }
+    if (reset) {
+        ctx->stats.readyCounter = 0;
+        ctx->stats.writeCounter = 0;
+        ctx->stats.compressedCounter = 0;
+    }
+    return ctx->compressionLevel;
+}
+
+static void* compressionThread(void* arg)
+{
+    adaptCCtx* ctx = (adaptCCtx*)arg;
+    unsigned currJob = 0;
+    for ( ; ; ) {
+        unsigned const currJobIndex = currJob % ctx->numJobs;
+        jobDescription* job = &ctx->jobs[currJobIndex];
+        DEBUGLOG(2, "compressionThread(): waiting on job ready\n");
+        pthread_mutex_lock(&ctx->jobReady_mutex);
+        while(currJob + 1 > ctx->jobReadyID) {
+            ctx->stats.waitReady++;
+            ctx->stats.readyCounter++;
+            DEBUGLOG(2, "waiting on job ready, nextJob: %u\n", currJob);
+            pthread_cond_wait(&ctx->jobReady_cond, &ctx->jobReady_mutex);
+        }
+        pthread_mutex_unlock(&ctx->jobReady_mutex);
+        DEBUGLOG(2, "compressionThread(): continuing after job ready\n");
+        /* compress the data */
+        {
+            unsigned const cLevel = adaptCompressionLevel(ctx);
+            DEBUGLOG(2, "cLevel used: %u\n", cLevel);
+            size_t const compressedSize = ZSTD_compressCCtx(ctx->cctx, job->dst.start, job->dst.size, job->src.start, job->src.size, cLevel);
+            if (ZSTD_isError(compressedSize)) {
+                ctx->threadError = 1;
+                DISPLAY("Error: something went wrong during compression: %s\n", ZSTD_getErrorName(compressedSize));
+                return arg;
+            }
+            job->compressedSize = compressedSize;
+        }
+        pthread_mutex_lock(&ctx->jobCompressed_mutex);
+        ctx->jobCompressedID++;
+        DEBUGLOG(2, "signaling for job %u\n", currJob);
+        pthread_cond_signal(&ctx->jobCompressed_cond);
```

Comment:
Yeah the performance of these things is very complex, especially with regards to fairness. I say we leave the signal inside the critical section because it is simpler, and as an added benefit it might perform better, though it probably doesn't matter.